inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp

```diff
@@ -1197,21 +1197,67 @@ bool MKLDNNGraph::InsertNode(MKLDNNNodePtr parent, MKLDNNNodePtr child, MKLDNNNo
 void MKLDNNGraph::EnforceBF16() {
     // Floating point parts of FP32 + INT8 or FP32 + BIN mixed precision models will be executed in BF16 precision
     // only if enforceBF16 flag was set manually because current performance is not good enough to enable it by default
-    if (implication(isQuantized(), config.manualEnforceBF16)) {
-        for (auto &node : graphNodes) {
-            if (node->getType() != Input && node->getType() != Output) {
-                for (size_t i = 0; i < node->getOriginalInputsNumber(); i++) {
-                    auto &parent = node->getParentEdgesAtPort(i)[0]->getParent();
-                    if (!(parent->getType() == Input && parent->isConstant()) &&       // exclude nodes after Constant Inputs
-                        !(parent->getType() == Input && node->getType() == Eltwise) && // exclude Eltwise after Input since it supports conversion to BF16
-                        node->getOriginalInputPrecisionAtPort(i) == Precision::FP32)
-                        node->setOriginalInputPrecisionAtPort(i, Precision::BF16);
-                }
+    if (!implication(isQuantized(), config.manualEnforceBF16))
+        return;
+    /* list of node types that must be forced to be executed in BF16 precision
+     * because of performance gains */
+    static const std::unordered_set<Type> significantNodes {
+        Convolution,
+        FullyConnected,
+        RNNCell,
+        RNNSeq,
+        MatMul,
+        ROIPooling,
+    };
 
-                for (size_t i = 0; i < node->getOriginalOutputsNumber(); i++) {
-                    if (node->getOriginalOutputPrecisionAtPort(i) == Precision::FP32)
-                        node->setOriginalOutputPrecisionAtPort(i, Precision::BF16);
-                }
+    auto shouldNotBeSkipped = [&](const MKLDNNNodePtr& node) {
+        return significantNodes.count(node->getType());
+    };
+
+    std::function<void(const MKLDNNNodePtr&, std::unordered_set<MKLDNNNodePtr>& skipNodes)> searchForNodesToSkip;
+    searchForNodesToSkip = [&](const MKLDNNNodePtr& node, std::unordered_set<MKLDNNNodePtr>& skipNodes) -> void {
+        for (const auto& edge : node->getParentEdges()) {
+            const auto& parent = edge.lock()->getParent();
+            if (shouldNotBeSkipped(parent)) // stop at significant nodes
```

Comment:
This lambda `shouldNotBeSkipped` can be inline. This way we can simplify the code. If this is not an obvious operation, just add an explanatory comment. In my opinion, it makes sense to use lambdas if we use such code more than once or need to pass a functor to some other operation.

Reply:
In general I would say the lambdas do make code harder to read in general, but considering amount of their usage in the plugin and the fact that every developer is used to use them, I think we can use them as a normal function, the same way that code is put into function no only in case of "reusability", but also to make it more declarative. So why to put the comment and do not use function if the function name itself is the comment.
Actually my comment was written before I introduced shouldNotBeSkipped lambda.

Reply:
Here you introduced lambda function instead of inline just one line of code, so people should scroll up to check what exactly this function does. Function extraction, with no doubts, very useful technic for complexity isolation and code reuse, but here is no excessive complexity and it is impossible to reuse this lambda in the outside, so you just introduced another function only for explanatory name, I think this reason is not good enough. M.b. this is a matter of taste, but we also have to use a common sense and make code easier for reading and understanding.

Reply:
I would say common sense is to jump to the definition and not to scroll. And I agree this it the matter of taste.

Reply:
The fewer the jumps, the easier it is to read.